### PR TITLE
fix(authenticator): fetch current user on routed apps after refresh

### PR DESCRIPTION
**Pull request by wlee221**
## Problem

Previously, `useAuthenticator` hook/composable/service required an instance of Authenticator in the DOM tree to initiate the machine and look for any current authenticated user in the local storage.

This caused problems when a user refreshed in an authenticated route. For example, imagine the following structure:

- `/` is the entry to the home page and has the `<Authenticator />` instance to enable sign in.
- `/my-account` is an _authenticated_ route that users can go after they signed in. This doesn't have `<Authenticator />` in DOM because they already signed in.

Whenever a user refreshed on `/my-account`, the page wouldn't have an `Authenticator` in the DOM to initiate the authMachine, and so auth contexts from the hooks returned undefined.

## Fix
This PR adjusts `useAuthenticator` hooks so that it first looks for `Auth.currentAuthenticatedUser`. If there are, it'll just transition directly to `authenticated` route and does not require an instance of `<Authenticator />` whatsoever. Otherwise, it would proceed to the `setup` state, in which it will wait for `Authenticator` to send the `INIT` event.

From the framework side, `<Authenticator />` will subscribe to authMachine changes. Once the machine gets to `setup` state and is ready for the `INIT` event, authenticator will send an INIT event with all the machine options.

## Remaining Problems

Note that there are still bugs we have to address for routed app use cases, namely #1581. This PR will at least correct the refresh behavior and remove the need for `Authenticator` in the DOM tree for machines to start.

#### Issue #, if available

Related to https://github.com/aws-amplify/amplify-ui/issues/1332, https://github.com/aws-amplify/amplify-ui/issues/1497

#### Description of how you validated changes
#1578 adds new examples and tests for refreshing. #1578 is to be merged into this PR once approved.

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are updated
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/ui/src/machines/authenticator/index.ts
Comment (wlee221) [inline packages/ui/src/machines/authenticator/index.ts:33]: `idle` state now starts with `Auth.currentAuthenticatedUser` call.

Knowing that, `idle` is a bad name (it's on me!) and should be named more explicitly. I scoped this out of this PR, because that would change our `route` names and I would rather deal with it in 3.0 changes.

File: packages/ui/src/machines/authenticator/index.ts
Comment (wlee221) [inline packages/ui/src/machines/authenticator/index.ts:48]: At this stage, we confirmed that no user has signed in yet. Now we wait for `Authenticator` to send `INIT` event to pass machine props like `initialState`.

File: packages/ui/src/machines/authenticator/index.ts
Comment (wlee221) [inline packages/ui/src/machines/authenticator/index.ts:56]: Once we have all the machine props ready, we apply those configs to get `loginMechanisms`, etc.

File: packages/angular/projects/ui-angular/src/lib/components/authenticator/components/authenticator/authenticator.component.ts
Comment (wlee221) [inline packages/angular/projects/ui-angular/src/lib/components/authenticator/components/authenticator/authenticator.component.ts:63]: Previously `authMachine` got stuck in `idle` state until an `Authenticator` in DOM sends an `INIT` event on mount. Now `authMachine` starts independently (see state machine changes below), and will explicitly ask for config from `Authenticator` once it's in `setup` state.

Once `Authenticator` sees that state is in `setup`, it'll send the usual `INIT` event.

File: packages/ui/src/machines/authenticator/index.ts
Comment (calebpollman) [inline packages/ui/src/machines/authenticator/index.ts:28]: Curious, can we just pass `defaultServices` here?

File: packages/vue/src/components/authenticator.vue
Comment (calebpollman) [inline packages/vue/src/components/authenticator.vue:85]: Just a couple questions:
- Should we unsubscribe after we have called `send`? (same question for the listener in Angular)
- What is the type of `newState`? May be missing context but it appears to be `Element` since we are calling `newState.matches` on ln 86, is that correct/expected?

File: packages/react/src/components/Authenticator/Authenticator.tsx
Comment (calebpollman) [inline packages/react/src/components/Authenticator/Authenticator.tsx:30]: Probably out of scope for this PR but what do you think of using `initial` as a route name in place of `setup`?

File: packages/vue/src/components/authenticator.vue
Comment (calebpollman) [inline packages/vue/src/components/authenticator.vue:87]: Nit: Think this can be combined with ln 86 in to one conditional statement

**Review by calebpollman — COMMENTED** [review]
Granted I am still learning how `xstate` and the `Authenticator` state machine work but LGTM 👍🏽. Left some questions

File: packages/ui/src/machines/authenticator/index.ts
Comment (wlee221) [reply packages/ui/src/machines/authenticator/index.ts:28]: Oh man, I did it again didn't I :D

Yep, you're right.

```suggestion
        services: defaultServices,
```

File: packages/vue/src/components/authenticator.vue
Comment (wlee221) [reply packages/vue/src/components/authenticator.vue:85]: > Should we unsubscribe after we have called send? 
Great point, yep `INIT` is a one time thing. Adjusting!

> What is the type of newState?
It's `XState.State` type, https://xstate.js.org/docs/guides/states.html#api. This is typed like so: 
image

File: packages/react/src/components/Authenticator/Authenticator.tsx
Comment (wlee221) [reply packages/react/src/components/Authenticator/Authenticator.tsx:30]: Yep, I don't like the name `setup` either, it's too generic and doesn't say much. Let's scope this out of the PR, this ties to #1581 as well.

File: packages/vue/src/components/authenticator.vue
Comment (wlee221) [reply packages/vue/src/components/authenticator.vue:87]: Yep :p

File: packages/ui/src/machines/authenticator/index.ts
Comment (ErikCH) [reply packages/ui/src/machines/authenticator/index.ts:56]: 👍

**Review by ErikCH — APPROVED** [review]
👍 

As long as the tests pass after you merge in your test PR

File: packages/vue/src/components/authenticator.vue
Comment (wlee221) [reply packages/vue/src/components/authenticator.vue]: Done in [924e906](https://github.com/aws-amplify/amplify-ui/pull/1580/commits/924e9066bdf25cdce7f9b60ded3bf6376df0a60c)

File: packages/vue/src/components/authenticator.vue
Comment (wlee221) [reply packages/vue/src/components/authenticator.vue]: unsubscription done in [c3a7daf](https://github.com/aws-amplify/amplify-ui/pull/1580/commits/c3a7dafcb61eda93c3642f4bd0aaa8c306bf5436) (vue) and [a72d2c3](https://github.com/aws-amplify/amplify-ui/pull/1580/commits/a72d2c314867ead5386d9ab870d62b652d2bcd1d) (angular).

**Review by calebpollman — APPROVED** [review]
LGTM 🚢  Thanks for addressing the feedback!

**Comment by wlee221** [comment]
fyi, I added 9 liner commit [b8c36bb](https://github.com/aws-amplify/amplify-ui/pull/1580/commits/b8c36bb60671864136d7b79228c47441fefff696) to fix a autoSignIn test failure. `autoSignIn` flow needed to be adjusted so that it goes to the correct state based on signUp result. This uses the same flow as before, just made more explicit.

I'll merge this tomorrow morning unless there are objections!

**Review by slaymance — APPROVED** [review]
Looks good to me!

**Pull request merged**
